### ISSUE TITLE
Memory error for downloaded image even though enough memory is available

### ISSUE THREAD

**Issue by mreichelt** [mixpanel/mixpanel-android#353]
TLDR: The error seems to be that the Mixpanel SDK uses `Runtime.getRuntime().freeMemory()` for checking if enough memory is available. This is wrong as the heap may expand dynamically, the documentation (for Android M) states:

```
    /**
     * Returns the number of bytes currently available on the heap without expanding the heap. See
     * {@link #totalMemory} for the heap's current size. When these bytes are exhausted, the heap
     * may expand. See {@link #maxMemory} for that limit.
     */
```

Exception was:

```
V/MixpanelAPI.DChecker: Can't load image https://URLREMOVED/image.jpg for a notification
                                                                     com.mixpanel.android.util.ImageStore$CantGetImageException: Do not have enough memory for the image
                                                                         at com.mixpanel.android.util.ImageStore.getImage(ImageStore.java:102)
                                                                         at com.mixpanel.android.mpmetrics.DecideChecker.getNotificationImage(DecideChecker.java:256)
                                                                         at com.mixpanel.android.mpmetrics.DecideChecker.runDecideCheck(DecideChecker.java:99)
                                                                         at com.mixpanel.android.mpmetrics.DecideChecker.runDecideChecks(DecideChecker.java:68)
                                                                         at com.mixpanel.android.mpmetrics.AnalyticsMessages$Worker$AnalyticsMessageHandler.handleMessage(AnalyticsMessages.java:263)
                                                                         at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                         at android.os.Looper.loop(Looper.java:148)
                                                                         at android.os.HandlerThread.run(HandlerThread.java:61)
```

The result is that currently no notifications can be shown because the SDK thinks the images are too large - which they are not.

**Comment by patedit** [comment]
You're totally right, `-Xms` and `-Xmx` don't necessarily need to be the same. Thanks for the PR, I just merged it on master.

**Comment by mreichelt** [comment]
@patedit do you have an ETA for the next bugfix version? This bug is currently blocking our release…
Thanks! :)

**Comment by patedit** [comment]
@mreichelt Sadly we are not planning to make a new release until at least two weeks. Do you think you could pull mixpanel dependency from github instead of maven? It'd be something like:

`compile "com.github.mixpanel:mixpanel-android:d947d5ee57"`

That is your commit and it's exactly version 4.8.6 plus your fix. Don't forget to add `maven { url "https://jitpack.io" }` in your project's repository dependencies.

If it's an inconvenience for you let me know and I'll make sure you have a release with your fix next Monday. Thanks!

**Comment by mreichelt** [comment]
@patedit Ok, I will do that. :)

I just thought this might be worth a minor (4.8.7) bugfix release, since one of *the* most important features of the SDK will simply fail to work - because the library won't show the notification. And not just for us, but for many people…

Besides, the parameters `-Xms` and `-Xmx` are something different - we are dealing with a dynamically changing heap size, not with startup parameters of a VM.

**Comment by patedit** [comment]
Those startup parameters specify the initial and maximum heap size. `totalMemory` and `maxMemory` correlate with those - hence my comment :) Dynamically changing the heap size or setting different values for `Xms` and `Xmx` are valid scenarios for this fix. Will keep an eye if someone else is having issues with notifications and will make a release if needed. Thanks @mreichelt !

**Comment by patedit** [comment]
@mreichelt You can find your fix on the latest release 4.8.7! Thanks again!

**Comment by mreichelt** [comment]
Sure thing :)
